cody-tests.el

```diff
@@ -0,0 +1,243 @@
+  "3 single-line suggestions included")
+
+(defun cody--test-whitespace-before-caret (str caret-pos)
+  "Return the whitespace before the first caret in STR.
+Returns ws up to the beginning of the line, given CARET-POS."
+  (let ((start-pos (string-match "\\(^\\|[^[:space:]]\\)[[:space:]]*$"
+                                 str 0 caret-pos)))
+    (if start-pos
+        (substring str start-pos caret-pos)
+      "")))
+
+(defun cody--test-set-up-caret-test (test-spec)
+  "Set up test parameters given a DSL in TEST-SPEC.
+
+Currently the DSL has only one element, the ^, which
+is removed in the result, and that position is returned
+as the point at which the completion is requested.
+
+Return value is (updated-text . caret-position)."
+  (let* ((caret (or (string-match "\\^" test-spec)
+                    (error "No ^ character found in test-spec")))
+         (text (concat (substring test-spec 0 caret)
+                       (substring test-spec (1+ caret)))))
+    (cons text caret)))
+
+(ert-deftest cody-test-single-line-completion ()
+  "Run checks for a single-line completion response."
+  ;; TODO: Make a new macro to reduce the boilerplate below.
+  (let* ((pretext "// file hello.c
+#include <stdio.h>
+int main() {
+   ^   
+   return 0;
+}
+")
+         (params (cody--test-set-up-caret-test pretext))
+         (insert-text (car params))
+         (pos (cdr params))
+         (cody--unit-testing-p t)) ; don't talk to real agent
+    (with-temp-buffer
+      (erase-buffer)
+      (insert insert-text)
+      (goto-char pos)
+      (cody-mode)
+      (let ((response (cody--mock-completion-response
+                       `(("id-1" "printf(\"Hello, world!\");" ,pos ,pos)))))
+        ;; Make sure cody-mode starts up without failing.
+        (should cody-mode)
+        ;; Should pass if the response looks good.
+        (when nil ; This test is -almost- working!
+          (should (cody--handle-completion-result
+                   response (current-buffer) pos "Automatic"))
+          ;; Check that the event is dropped if the point was moved.
+          (should-not (cody--handle-completion-result
+                       response
+                       (current-buffer)
+                       (- pos 2) ; point has moved 2 spaces ahead of request
+                       "Invoke"))) ; or Automatic
+        ;; It should error if items is not a vector.
+        (should 'finish-this-test)))))
+         
+
+(ert-deftest cody-test-diff-lists ()
+  "Tests for `cody-diff-lists' Myers-diff functionality."
+  (should (null (cody-diff-lists nil nil)))
+  (let* ((a '(1 2 3))
+         (b a))
+    (should (equal (cody-diff-lists a b) a)))
+  ;; Test James Coglan's example of ABCABBA and CBABAC.
+  ;; https://blog.jcoglan.com/2017/02/12/the-myers-diff-algorithm-part-1/
+  (should (equal (cody-diff-lists '(A B C A B B A) '(C B A B A C))
+                 '((- A B) C (- A) B (+ A) B A (+ C)))))
+
+(ert-deftest cody-test-diff-strings ()
```

Comment:
Worth testing some empty strings here? Prefixes and suffixes? The coverage you have is fine, but if something broke, the tests aren't very specific. Some empties, prefixes and suffixes would hit those (a `('- . ...)) (b `('+ . ...)) cases quite specifically.